domain/domain.go

```diff
@@ -222,37 +220,60 @@ func (do *Domain) GetScope(status string) variable.ScopeFlag {
 	return variable.DefaultScopeFlag
 }
 
-func (do *Domain) tryReload() {
-	// If we don't have update the schema for a long time > lease, we must force reloading it.
-	// Although we try to reload schema every lease time in a goroutine, sometimes it may not run accurately.
-	// e.g., the machine has a very high load, running the ticker is delayed.
-	last := atomic.LoadInt64(&do.lastLeaseTS)
-	lease := do.ddl.GetLease()
+func (do *Domain) mockReloadFail() error {
+	err := kv.RunInNewTxn(do.store, false, func(txn kv.Transaction) error {
+		do.SchemaValidity.setLastFailedTS(txn.StartTS())
+		return nil
+	})
+	if err != nil {
+		log.Errorf("mock reload failed err:%v", err)
+		return errors.Trace(err)
+	}
+	return errors.New("mock reload failed")
+}
 
-	// if lease is 0, we use the local store, so no need to reload.
-	if lease > 0 && time.Now().UnixNano()-last > lease.Nanoseconds() {
-		do.MustReload()
+func (do *Domain) doReload(exit *int32) error {
+	var err error
+	for {
+		var ver kv.Version
+		ver, err = do.store.CurrentVersion()
+		if err == nil {
+			schemaVersion := int64(0)
+			oldInfoSchema := do.infoHandle.Get()
+			if oldInfoSchema != nil {
+				schemaVersion = oldInfoSchema.SchemaMetaVersion()
+			}
+			err = do.loadInfoSchema(do.infoHandle, schemaVersion, ver.Ver)
+		}
+		if err == nil {
+			atomic.StoreInt64(&do.lastLeaseTS, time.Now().UnixNano())
+			break
+		}
+		do.SchemaValidity.setLastFailedTS(ver.Ver)
+		log.Errorf("[ddl] load schema err %v, retry again", errors.ErrorStack(err))
+		if atomic.LoadInt32(exit) == 1 {
+			break
+		}
+		// TODO: Use a backoff algorithm.
+		time.Sleep(doReloadSleepTime)
+		continue
 	}
+	return err
 }
 
+const doReloadSleepTime = 500 * time.Millisecond
+const mustReloadSleepTime = 1 * time.Second
+
 var defaultMinReloadTimeout = 20 * time.Second
 
 // Reload reloads InfoSchema.
-func (do *Domain) Reload() error {
+func (do *Domain) reload() error {
 	// for test
 	if do.SchemaValidity.MockReloadFailed {
```

Comment:
mixing the test code logic would hurt readability, can you find better way to do it? 
such as extract to an interface and wrap the mocked code upon the real implementation